File: 백준/Gold/3079. 입국심사/입국심사.py
from sys import stdin

input = stdin.readline

n, m = map(int, input().split())

enterances = [int(input()) for _ in range(n)]

INF = int(1e18)
left, right = 0, INF
answer = INF

while left <= right:
    mid = (left + right) // 2 # 걸리는 시간

    total = 0
    for enterance in enterances:
        total += (mid // enterance)

    if total >= m:
        right = mid - 1
        answer = min(answer, mid)
    else:
        left = mid + 1

print(answer)

File: 백준/Gold/3079. 입국심사/입국심사.java
import java.io.*;
import java.util.*;

class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String[] split = br.readLine().split(" ");
        int n = Integer.parseInt(split[0]);
        int m = Integer.parseInt(split[1]);

        int[] gate = new int[n];
        for(int i=0 ; i<n ; i++){
            gate[i] = Integer.parseInt(br.readLine());
        }

        long left = 0, right = (long)1e18;

        long answer = right;
        while(left <= right){
            long mid = (left + right) / 2;

            long count = 0;
            for(int i=0 ; i<n ; i++){
                if(count + (mid / gate[i]) < 0)break;
                count += (mid /gate[i]);
            }

            if(count >= m){
                right = mid - 1;
                answer = Math.min(answer, mid);
            }
            else{
                left = mid + 1;
            }
        }

        bw.write(String.valueOf(answer));
        bw.flush();
        bw.close();
        br.close();
    }
}